### PR TITLE
Warn user if using e2ee and vfs together

### PR DESCRIPTION
**Pull request by FlexW**
Signed-off-by: Felix Weilbach <felix.weilbach@nextcloud.com>

### CODE REVIEW

File: src/gui/accountsettings.cpp [hunk 1]
```diff
@@ -62,6 +62,27 @@
...
+
+bool showEnableE2eeWithVirtualFilesWarningDialog()
+{
+    QMessageBox e2eeWithVirtualFilesWarningMsgBox(QMessageBox::Warning,
+        "End-to-End Encryption with Virtual Files",
+        "You seem to have the Virtual Files feature enabled on this folder. Virtual"
+        " Files and End-to-End Encryption work not well together at the moment. You"
+        " can allow End-to-End Encryption on this folder, but please keep in mind that"
+        " you need to mark the encrypted folder with \"Make always available locally\" before"
+        " creating or modifying files in it.");
```
Comment (er-vin) [inline src/gui/accountsettings.cpp:74]: I'm surprised about that strong statement. Isn't it *only* the implicit hydration which fails? Everything else should work fine AFAIK (even adding files in an otherwise unavailable E2EE folder).

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp [hunk 2]
```diff
@@ -308,6 +309,15 @@ void VfsCfApi::requestHydration(const QString &requestId, const QString &path)
...
     if (record._isE2eEncrypted || !record._e2eMangledName.isEmpty()) {
         qCInfo(lcCfApi) << "Couldn't hydrate, the file is E2EE this is not supported";
+
+        QMessageBox e2eeFileDownloadRequestWarningMsgBox(QMessageBox::Warning,
+            "Download of End-to-End encrypted file failed",
+            "It seems that you are trying to download a virtual file that"
+            " is End-to-End encrypted. Implicitly downloading such files is not"
+            " supported at the moment. To workaround that issue, go to the"
+            " settings and mark the encrypted folder with \"Make always available locally\".");
+        e2eeFileDownloadRequestWarningMsgBox.exec();
```
Comment (er-vin) [inline src/libsync/vfs/cfapi/vfs_cfapi.cpp:319]: Better use one of the static methods if you do this.

https://doc.qt.io/qt-5/qmessagebox.html#warning

File: src/gui/accountsettings.cpp (hunk 1 above)
Comment (FlexW) [reply src/gui/accountsettings.cpp:74]: Yes, I expected too that it would work, but I tried to create files in a e2ee folder that is not marked as always available, and it failed with an assert. These were the captured logs in case you are interessted:

```
2021-04-22 12:26:10:405 [ info nextcloud.sync.credentials.webflow ]:	request finished
2021-04-22 12:26:10:405 [ info nextcloud.sync.networkjob.put ]:	PUT of "http://localhost:8080/remote.php/dav/files/admin/myencfolder/570887b6bfd74bd88febd1a07b44a782" FINISHED WITH STATUS "OK" QVariant(int, 201) QVariant(QString, "Created")
2021-04-22 12:26:10:405 [ critical nextcloud.sync.csync.vio_local ]:	CreateFileW failed on "\\\\?\\C:\\Users\\felix\\workspace\\desktop\\build\\myfiles\\myencfolder\\New Text Document.txt"
2021-04-22 12:26:10:405 [ warning nextcloud.sync.filesystem ]:	Could not get modification time for "C:/Users/felix/workspace/desktop/build/myfiles/myencfolder/New Text Document.txt" with csync, using QFileInfo
2021-04-22 12:26:10:405 [ info nextcloud.sync.filesystem ]:	File "C:/Users/felix/workspace/desktop/build/myfiles/myencfolder/New Text Document.txt" has changed: size:  0 <-> 0 , mtime:  1619087167 <-> -3600
2021-04-22 12:26:10:405 [ critical nextcloud.sync.vfs.cfapi.wrapper ]:	Could not CreateFile for longpath: "\\\\?\\C:\\Users\\felix\\workspace\\desktop\\build\\myfiles\\myencfolder\\New Text Document.txt" with error: 2
2021-04-22 12:26:10:405 [ fatal default ]:	ASSERT: "handle" in file ..\src\libsync\vfs\cfapi\cfapiwrapper.cpp, line 576
```

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 2 above)
Comment (FlexW) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:319]: But then I have to pass a parent widget, and I don't have one in this place.

File: src/gui/accountsettings.cpp (hunk 1 above)
Comment (FlexW) [reply src/gui/accountsettings.cpp:74]: I tried it again and now I was able to create e2ee files, even if the folder is not set to make available offline. Maybe the other error came frome something else. Anyway I think it's a good idea to not create e2ee folders when using vfs.

Comment (er-vin) [reply src/gui/accountsettings.cpp:74]: Not getting the modtime of a file you created is indeed odd. Seems vfs related but I somehow doubt this is E2EE related. Could be wrong of course, can't beat having a debugger attached. ;-)

Comment (er-vin) [reply src/gui/accountsettings.cpp:74]: I'd tone it down in any case. Right now it's going to scare away users thinking it'll eat their puppies.

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 2 above)
Comment (er-vin) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:319]: Just pass nullptr :-)

You'd have to pass one with the instance too, you just benefit from the `= nullptr` on that parameter.

Comment (FlexW) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:321]: Okay. While rereading the documentation, I have seen that the explaining text should be set with `setInformativeText()`. Seems like that is not possible with the static methods.

File: src/gui/accountsettings.cpp (hunk 1 above)
Comment (FlexW) [reply src/gui/accountsettings.cpp]: Do you like it now more?

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 2 above)
Comment (er-vin) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:321]: Well, I'd expect what you pass to setText to be the title and what you pass to setInformativeText to be the text.

File: src/gui/accountsettings.cpp [hunk 3]
```diff
@@ -62,6 +62,28 @@
...
+{
+    QMessageBox e2eeWithVirtualFilesWarningMsgBox;
+    e2eeWithVirtualFilesWarningMsgBox.setText("End-to-End Encryption with Virtual Files");
+    e2eeWithVirtualFilesWarningMsgBox.setInformativeText("You seem to have the Virtual Files feature enabled on this folder. At "
+                                                         " the moment, it is not possible to implicit downloading virtual files that are "
+                                                         "End-to-End encrypted. To get the best experience with Virtual Files and"
+                                                         " End-to-End Encryption, make sure the encrypted folder is marked with"
+                                                         " \"Make always available locally\".");
+    e2eeWithVirtualFilesWarningMsgBox.setIcon(QMessageBox::Warning);
+    const auto dontEncryptButton = e2eeWithVirtualFilesWarningMsgBox.addButton("Don't encrypt this folder", QMessageBox::AcceptRole);
```
Comment (er-vin) [inline src/gui/accountsettings.cpp:76]: It's more common to override text of say... ok and cancel and then rely on if it was accepted or not. This avoids the unreachable and you can return a simple condition over the exec() value like for a regular dialog.

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 2 above)
Comment (FlexW) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:321]: But as far as I understand the documentation, the title should be just the application name to be conform with Windows and macOS human interface guidelines.

Comment (er-vin) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:321]: Not uncommon to have more or something else in there, but OK, maybe. I admit I didn't check their HIG in a while.

File: src/gui/accountsettings.cpp [hunk 4]
```diff
@@ -62,6 +62,28 @@
 
 namespace {
 constexpr auto propertyFolderInfo = "folderInfo";
+
+bool showEnableE2eeWithVirtualFilesWarningDialog()
+{
+    QMessageBox e2eeWithVirtualFilesWarningMsgBox;
+    e2eeWithVirtualFilesWarningMsgBox.setText("End-to-End Encryption with Virtual Files");
```
Comment (allexzander) [inline src/gui/accountsettings.cpp:69]: @FlexW Are the QMessageBox texts always considered translatable? I just didn't find anything about that in the QMessageBox's documentation page, while I don't see a `tr()` wrapper for the string you pass here either.

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 2 above)
Comment (allexzander) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:321]: If this is really the case and the static method suggested by @er-vin indeed is incomplete for this scenario, we may consider implementing our own static method that `does it all` with QMessageBox. Just to create tools we can reuse in these 2 places where the message box is needed, and, also in the future. Just a suggestion of possible compromise.

File: src/gui/accountsettings.cpp (hunk 4 above)
Comment (er-vin) [reply src/gui/accountsettings.cpp:69]: D'oh, yes, stupid me. It's missing tr() everywhere here.

Comment (FlexW) [reply src/gui/accountsettings.cpp:69]: thanks @allexzander. I forgot that completely.

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 2 above)
Comment (FlexW) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:321]: That is a good idea @allexzander , but I'm not sure where to create it. One message box is part libsync, the other of gui. Then there is the common folder, but in this folder but there is a comment that it should compile without gui. Any suggestions?

File: src/gui/accountsettings.cpp [hunk 5]
```diff
@@ -55,13 +55,34 @@
...
 namespace {
 constexpr auto propertyFolderInfo = "folderInfo";
+
+bool showEnableE2eeWithVirtualFilesWarningDialog(QObject *parent)
+{
+    Q_ASSERT(parent);
+    QMessageBox e2eeWithVirtualFilesWarningMsgBox;
+    e2eeWithVirtualFilesWarningMsgBox.setText(parent->tr("End-to-End Encryption with Virtual Files"));
+    e2eeWithVirtualFilesWarningMsgBox.setInformativeText(parent->tr("You seem to have the Virtual Files feature enabled on this folder. At "
+                                                                    " the moment, it is not possible to implicit downloading virtual files that are "
```
Comment (camilasan) [inline src/gui/accountsettings.cpp:70]: I think it is "to implicit download" not "to implicit downloading".

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp [hunk 6]
```diff
@@ -308,6 +309,17 @@ void VfsCfApi::requestHydration(const QString &requestId, const QString &path)
     // of the placeholder with decrypted data
     if (record._isE2eEncrypted || !record._e2eMangledName.isEmpty()) {
         qCInfo(lcCfApi) << "Couldn't hydrate, the file is E2EE this is not supported";
+
+        QMessageBox e2eeFileDownloadRequestWarningMsgBox;
+        e2eeFileDownloadRequestWarningMsgBox.setText(tr("Download of End-to-End encrypted file failed"));
+        e2eeFileDownloadRequestWarningMsgBox.setInformativeText(tr("It seems that you are trying to download a virtual file that"
+                                                                   " is End-to-End encrypted. Implicitly downloading such files is not"
+                                                                   " supported at the moment. To workaround that issue, go to the"
```
Comment (camilasan) [inline src/libsync/vfs/cfapi/vfs_cfapi.cpp:317]: I think it is "to workaround _this_ issue" not "to workaround _that_ issue".

**Review by camilasan — CHANGES_REQUESTED** [review]

File: src/gui/accountsettings.cpp [hunk 7]
```diff
@@ -55,13 +55,34 @@
...
+    e2eeWithVirtualFilesWarningMsgBox.setText(parent->tr("End-to-End Encryption with Virtual Files"));
+    e2eeWithVirtualFilesWarningMsgBox.setInformativeText(parent->tr("You seem to have the Virtual Files feature enabled on this folder. At "
+                                                                    " the moment, it is not possible to implicit download virtual files that are "
+                                                                    "End-to-End encrypted. To get the best experience with Virtual Files and"
+                                                                    " End-to-End Encryption, make sure the encrypted folder is marked with"
+                                                                    " \"Make always available locally\"."));
+    e2eeWithVirtualFilesWarningMsgBox.setIcon(QMessageBox::Warning);
+    const auto dontEncryptButton = e2eeWithVirtualFilesWarningMsgBox.button(QMessageBox::StandardButton::Ok);
+    const auto encryptButton = e2eeWithVirtualFilesWarningMsgBox.button(QMessageBox::StandardButton::Cancel);
+    dontEncryptButton->setText(parent->tr("Encrypt this folder"));
```
Comment (allexzander) [inline src/gui/accountsettings.cpp:77]: @FlexW Looks like there is a mistake with button texts. Both have the same text.

File: src/gui/accountsettings.cpp [hunk 8]
```diff
@@ -55,13 +55,34 @@
...
+    encryptButton->setText(parent->tr("Encrypt this folder"));
+
+    if (e2eeWithVirtualFilesWarningMsgBox.clickedButton() == dontEncryptButton) {
```
Comment (allexzander) [inline src/gui/accountsettings.cpp:80]: @FlexW You can shorten this to `return e2eeWithVirtualFilesWarningMsgBox.clickedButton() != dontEncryptButton`

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 2 above)
Comment (allexzander) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:321]: @FlexW Within the `guiutility.h` for example.

File: src/gui/accountsettings.cpp (hunk 7 above)
Comment (FlexW) [reply src/gui/accountsettings.cpp:77]: thanks

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 2 above)
Comment (FlexW) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:321]: @allexzander but then it not usable for one of the dialogs, because one of the dialogs is in libsync. maybe the architecture can be changed, that both dialogs get created within gui.

Comment (allexzander) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp:321]: > @allexzander but then it not usable for one of the dialogs, because one of the dialogs is in libsync. maybe the architecture can be changed, that both dialogs get created within gui.

@FlexW Alrighty then. I wouldn't hurry with this helper function in this case.

File: src/gui/accountsettings.cpp (hunk 5 above)
Comment (allexzander) [reply src/gui/accountsettings.cpp]: > I think it is "to implicit download" not "to implicit downloading".

@FlexW @camilasan I'd add one more improvement to Camila's suggestion. Should have **ly** at the end:
implicit**ly**

**Review by allexzander — APPROVED** [review]

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 6 above)
Comment (er-vin) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp]: Depends on the distance of the reader from the issue I guess :clown_face:

Sorry couldn't resist, the this/that distinction in English is so messy. :laughing:

File: src/gui/accountsettings.cpp [hunk 9]
```diff
@@ -55,13 +55,30 @@
...
 #include "account.h"
 
 namespace {
 constexpr auto propertyFolderInfo = "folderInfo";
+
+bool showEnableE2eeWithVirtualFilesWarningDialog(QObject *parent)
+{
+    Q_ASSERT(parent);
+    QMessageBox e2eeWithVirtualFilesWarningMsgBox;
+    e2eeWithVirtualFilesWarningMsgBox.setText(parent->tr("End-to-End Encryption with Virtual Files"));
```
Comment (er-vin) [inline src/gui/accountsettings.cpp:68]: Better either use `AccountSettings::tr()` (this is static) or `QCoreApplication::translate()`. Please don't pass an (in fact) unused instance.

https://doc.qt.io/qt-5/i18n-source-translation.html#using-tr-for-all-literal-text

File: src/libsync/vfs/cfapi/vfs_cfapi.cpp (hunk 6 above)
Comment (allexzander) [reply src/libsync/vfs/cfapi/vfs_cfapi.cpp]: > Depends on the distance of the reader from the issue I guess 🤡
> 
> Sorry couldn't resist, the this/that distinction in English is so messy. 😆

"That's not an issue. THAT's an issue!" (c) Altered quote from some old movie 🤡

File: src/gui/accountsettings.cpp (hunk 9 above)
Comment (allexzander) [reply src/gui/accountsettings.cpp:68]: Good catch indeed.

File: src/gui/accountsettings.cpp (hunk 3 above)
Comment (FlexW) [reply src/gui/accountsettings.cpp]: Is it now better?

File: src/gui/accountsettings.cpp (hunk 9 above)
Comment (FlexW) [reply src/gui/accountsettings.cpp]: Thanks. I used `QObject::tr()` now. That should work also I think? Wanted to make it more general and not dependent on `AccountSettings`. `QCoreApplication::translate()` is more complicated to call.

**Comment by FlexW** [comment]
@camilasan are you fine with it now?

**Review by mgallien — APPROVED** [review]
thanks

**Review by camilasan — APPROVED** [review]
